### PR TITLE
SNOW-1064310: local test implement session fileoperation get

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-1064310, add file operation get support

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_stage_registry.py [hunk 1]
```diff
@@ -179,6 +207,177 @@ def upload_stream(
...
+                        file_name,
+                        file_size,
+                        "DOWNLOADED",
+                        "",
+                    ],
+                )
+            )
+        return result_df
+
+    def read_file(
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_stage_registry.py:275]: this method is moved from the deleted _file_operation.py to consolidate file operations in the stage registry module

**Review by sfc-gh-jrose — APPROVED** [review]
Generally looks good to me.

File: tests/integ/scala/test_file_operation_suite.py [hunk 2]
```diff
@@ -460,98 +460,135 @@ def test_put_stream_negative(session, temp_stage, path1, local_testing_mode):
...
         assert len(results) == len(results_with_statement_params) == 1
-        assert (
-            results[0].file
-            == results_with_statement_params[0].file
-            == f"{os.path.basename(path1)}.gz"
-        )
-        assert results[0].size in (54, 55) and results_with_statement_params[
+        assert results[0].file == results_with_statement_params[0].file == file_name
+        # 10 is for local testing, 54, 55 is for non-local testing
+        assert results[0].size in (10, 54, 55) and results_with_statement_params[
```
Comment (sfc-gh-stan) [inline tests/integ/scala/test_file_operation_suite.py:495]: Let's use `if local_testing_mode` pattern, I think it's always better to be explicit.

File: src/snowflake/snowpark/mock/_stage_registry.py [hunk 3]
```diff
@@ -195,15 +225,192 @@ def upload_stream(
...
+                column_series = ColumnEmulator(
+                    data=None, dtype=object, name=column_name
+                )
+                column_series.sf_type = ColumnType(
+                    schema[i].datatype, schema[i].nullable
+                )
+                result_df[column_name] = column_series
+                result_df_sf_types[column_name] = column_series.sf_type
+                if type(column_series.sf_type.datatype) not in CONVERT_MAP:
+                    _logger.warning(
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_stage_registry.py:348]: Question: what is this case for?

File: src/snowflake/snowpark/mock/_stage_registry.py [hunk 4]
```diff
@@ -241,3 +448,43 @@ def upload_stream(
...
         )
+
+    def get(
+        self,
+        stage_location: str,
+        target_directory: str,
+        options: Dict[str, str] = None,
+    ):
+        if not stage_location.startswith("@"):
+            raise SnowparkSQLException("SQL compilation error")
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_stage_registry.py:459]: We discussed the error behavior last week, let's try to change all SQL specific errors to something more debuggable.

File: src/snowflake/snowpark/mock/_stage_registry.py [hunk 5]
```diff
@@ -241,3 +448,43 @@ def upload_stream(
...
+
+    def read_file(
+        self,
+        stage_location,
+        format: str,
+        schema: List[Attribute],
+        analyzer: "MockAnalyzer",
+        options: Dict[str, str],
+    ):
+        if not stage_location.startswith("@"):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_stage_registry.py:478]: ditto

**Review by sfc-gh-stan — APPROVED** [review]
LGTM with a few NIT-ish comments.

File: src/snowflake/snowpark/mock/_stage_registry.py [hunk 6]
```diff
@@ -179,6 +207,177 @@ def upload_stream(
...
+    def get_file(
+        self,
+        stage_location: str,
+        target_directory: str,
+        options: Dict[str, str] = None,
+    ) -> TableEmulator:
+        if target_directory.startswith("'file://"):
+            target_directory = target_directory[
+                len("'file://") : -1
+            ]  # skip normalized prefix `file:// and suffix `
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_stage_registry.py:248]: this is to drop the tailing backtick "`", the local_file_name here is already normalized so `/` is handled.
but good to learn it, thanks

File: src/snowflake/snowpark/mock/_stage_registry.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_stage_registry.py:491]: updated to use more explicit error message

File: src/snowflake/snowpark/mock/_stage_registry.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_stage_registry.py]: updated to use more explicit error message

File: tests/integ/scala/test_file_operation_suite.py (hunk 2 above)
Comment (sfc-gh-aling) [reply tests/integ/scala/test_file_operation_suite.py]: thanks for catching this, updated

File: src/snowflake/snowpark/mock/_stage_registry.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_stage_registry.py]: users can specify schema when reading csv file, and the schema(types) decides how to we want do the data conversion (from string to xxx type) when reading the csv file into a dataframe.

CONVERT_MAP defined the types and conversion method we support when reading data.

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_stage_registry.py]: also good catch that we should use the oob telemetry to log the event.

**Pull request merged**